### ISSUE TITLE
load qtquick slideshow ahead of time (+ possibly send a start signal)

### ISSUE THREAD

**Issue by apachelogger** [calamares/calamares#1152]
**Describe the bug**
Currently (on git master) it looks like the qtquick scene is only loaded when the view actually goes to the exec step. This has the fairly jarring side effect that there may be a very noticeable delay before something actually shows up.

**To Reproduce**
1. have a slideshow (with a large-ish image ideally so it takes a while to load)
2. do an installation
3. observe noticeable time of nothing-ness upon reaching the slideshow page

**Expected behavior**
When moving to the slideshow page the slideshow should be rendered and started immediately.

**Screenshots and Logs**
Video of a case where this happens http://people.ubuntu.com/~apachelogger/screencasts/vokoscreen-2019-05-20_15-09-40.mkv

**Additional context**
Ideally two things should happen I suppose:

1. The qml scene should be loaded ahead of time. This doesn't necessarily have to be directly at start up of calamares, but maybe after a 500ms timer expires or something. So it doesn't block startup and yet all components have been constructed before they are needed.
2. The slideshow API should be extended with a `start` signal sent by calamares when the view step is reached. Slideshows should then activate their auto progress timer (if they have one) after receiving start. Possibly acting on `onVisibleChanged` would be enough as well, but perhaps an explicit start API would be harder to miss when writing a slideshow implementation.

**Comment by adriaandegroot** [comment]
QML loading (and compilation) can be done asynchronously, so I've done that.

The QML component is still created at the moment the slideshow page is reached, but at least the engine initialization and whatnot is already done. Adding explicit start/stop signals to the QML itself would probably be most convenient, but that means modifying all the slideshows out there to respect that; the current example slideshows all start a timer at `onComplete`, which means you'd start the timer at Calamares startup, and hit an essentially random "first page" of the slideshow.

The solution picked now requires no changes to any existing QML.

**Comment by adriaandegroot** [comment]
Reopening. It's not part of 3.2.9, and 3.2.10 is partitioning-related, so it'll happen later, after more careful testing **and** figuring out a good API for this (from the slideshow side).

**Comment by apachelogger** [comment]
Quick mind dump based on my musing elsewhere:

This really needs a proper framework for synchronization of start/stop behavior and proper pre-loading of the scene (as in: at or shortly after application start). All other approaches will suck to varying degrees.

The situation is complicated by the facts that 
a) the slideshows may be arbitrary QML unrelated to the Presentation framework (which IMO should be revisited in a future incompatibility change, there really need to be some boundaries or set expectations)
b) current Presentation expects that once the root Component is complete it is shown and may trigger any timers (this of course goes against fully pre-loading the scene).

My proposed solution would be this:
- new C++ item subclassing QQuickItem as PresentationBase or ShowBase or some such name
  - new signals `startPresentation` and `stopPresentation` (there probably is no use case for stop but whatever); or just a bool property `running`.
  - overriding the `connectNotify` function to track when something is connected to the signals
- the current Presentation.qml instead of deriving an Item derives PresentationBase
- all existing Presentation-based slideshows can now `onStartPresentation` and `onStopPresentation` to use the new start/stop framework
- calamares on startup fully initializes the qquickview
  - get https://doc.qt.io/qt-5/qquickview.html#rootObject
  - qobject_cast to PresentationBase
  - if the object is a PresentationBase and its start/stop were connected to: keep the view around and use it once the slideshow page is used for display
    - call start and stop when showing hiding the slideshow page respectively
  - else: throw the view away and initialize it again when showing the slideshow page (i.e. behave exactly as is the case right now)

```
Presentation {
  onStart: timer.running = true
  onStop: timer.running = false
}
```

This makes everything work splendidly. Old shows regardless of type have zero behavior change. When they are ported to PresentationBase as root item they can use start/stop. Calamares automatically can detect whether or not start/stop is supported by a given slideshow.

Long term everything should use PresentationBase and it should become an requirement for all slideshow implementations to have the base as their root. Within that root they may still choose to do whatever, since the base is a regular Item it doesn't have any preconceived notions about the world.

An alternative, but similar approach is one without inheritance. Instead of changing the root object calamares could set a `contextProperty` to expose a QObject which the slideshow then connects to for start/stop. This is a bit garbage to use from the QML side though.

```
Presentation {
  Connections {
    target: fancyContextProperty
    onStart: timer.running = true
    onStop: timer.running = false
  }
}
```

An in-between sort of approach would be to have a StartStopItem written in C++ that is made available to QML via `qmlRegisterSingletonType`. This means we have a singleton to talk to on the C++ side without casting and it's still very idiomatic on the QML side

```
Presentation {
  StartStopItem {
    onStart: timer.running = true
    onStop: timer.running = false
  }
}
```

Another point of note is that signal detection isn't technically necessary. One could just as well expect the branding.conf to set `preload: true` on the slideshow, which would carry the expectation that either timing doesn't matter or start/stop was implemented. This would possibly mean less code overall since cala loads the branding.conf anyway, so it can decide then and there if the view can be preloaded (supports start/stop) or not. Certainly less faffy than listening to connectNotify.

If you have more questions on any of this, please ask.
